nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_dataset.py

```diff
@@ -618,17 +618,28 @@ def collate_fn(self, batch):
 
         position_ids: List[List[int]] = []
         cu_seqlens: List[List[int]] = []
+        cu_seqlens_unpadded: List[List[int]] = []
         for item in batch:
             position_ids.append([])
             cu_seqlens.append([0])
+            cu_seqlens_unpadded.append([0])
             seqlens = np.array(item['seq_boundaries'][1:]) - np.array(item['seq_boundaries'][:-1])
             for l in seqlens:
                 # length minus 1 because input_ids is truncated by 1 for labels
                 position_ids[-1].extend(list(range(l - 1)))
                 cu_seqlens[-1].append(cu_seqlens[-1][-1] + l - 1)
-            # set last seq to the max seq len because rope and attn kernels expect no padding
+
+            # the last seq needs to be the max seq len because rope and attn kernels expect no padding
             cu_seqlens[-1][-1] = max_length
 
+            for i in range(len(item['seq_boundaries']) - 1):
+                seqlen_unpadded = np.where(
+                    np.array(item['input_ids'][item['seq_boundaries'][i] : item['seq_boundaries'][i + 1] - 1])
+                    == self.tokenizer.eos_id
+                )
+                seqlen_unpadded = seqlen_unpadded[0][1] if seqlen_unpadded[0].size > 1 else seqlen_unpadded[0][0]
```

Comment:
what does this line of code mean? why not always `seqlen_unpadded[0][0]`?

Reply:
The padding added in the data preparation code is EOS token. We need to make sure that one EOS token is not counted as padding. Thus it's not always `seqlen_unpadded[0][0]`.

Reply:
I think I do not know what is `seqlen_unpadded[0].size`? why do we need to check if it's bigger than 1? could you please elaborate this? thanks

Reply:
I have pushed a new commit renaming the some variables and adding some comments. Hopefully it will be more clear.

Reply:
Thanks, I am just thinking all tokens are padded at the end, so if `eos_idx[0].size > 1`, isn't `eos_idx[0][1] == eos_idx[0][0] + 1`? If yes, shouldn't we just do `seqlen_unpadded = eos_idx[0][0] + 1`?

Reply:
Yes, I just updated that.